src/Orleans.Core/Runtime/ThreadPerTaskScheduler.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime
+{
+    /// <summary>Provides a task scheduler that dedicates a thread per task.</summary>
+    public class ThreadPerTaskScheduler : TaskScheduler
+    {
+        private readonly Func<Task, string> threadNameProvider;
+
+        public ThreadPerTaskScheduler(Func<Task, string> threadProvider)
+        {
+            this.threadNameProvider = threadProvider;
+        }
+
+        /// <summary>Gets the tasks currently scheduled to this scheduler.</summary>
+        /// <remarks>This will always return an empty enumerable, as tasks are launched as soon as they're queued.</remarks>
+        protected override IEnumerable<Task> GetScheduledTasks()
+        {
+            return Enumerable.Empty<Task>();
+        }
+
+        /// <summary>Starts a new thread to process the provided task.</summary>
+        /// <param name="task">The task to be executed.</param>
+        protected override void QueueTask(Task task)
+        {
+            new Thread(() => TryExecuteTask(task))
```

Comment:
Don't we need some method for joining/aborting/disposing these threads during shutdown? Eg by making this & ExecutorService disposable, so that ExecutorService disposes this. That way, both will be disposed when the container is disposed.

Reply:
I tried to introduce as little additional changes as possible. The `AsynchAgent`s threads weren't being joined on dispose before, and I I didn't see the need in it here.

Reply:
After thinking about it more, I agree with you. Current code uses a cooperative cancellation approach and it's best if we continue doing that.